core/src/main/scala/com/microsoft/azure/synapse/ml/causal/BaseDiffInDiffEstimator.scala

```diff
@@ -0,0 +1,183 @@
+      .map(new LinearRegression().setWeightCol)
+      .getOrElse(new LinearRegression())
+
+    regression
+      .setFeaturesCol(featuresCol)
+      .setLabelCol(getOutcomeCol)
+      .setFitIntercept(fitIntercept)
+      .setLoss("squaredError")
+      .setRegParam(1E-10)
+
+    assembler.transform _ andThen regression.fit apply df
+  }
+}
+
+case class DiffInDiffSummary(treatmentEffect: Double, standardError: Double,
+                             timeWeights: Option[DVector] = None,
+                             timeIntercept: Option[Double] = None,
+                             timeRMSE: Option[Double] = None,
+                             unitWeights: Option[DVector] = None,
+                             unitIntercept: Option[Double] = None,
+                             unitRMSE: Option[Double] = None,
+                             zeta: Option[Double] = None,
+                             lossHistoryTimeWeights: Option[List[Double]] = None,
+                             lossHistoryUnitWeights: Option[List[Double]] = None) {
+  import scala.collection.JavaConverters._
+
+  def getLossHistoryTimeWeightsJava: Option[util.List[Double]] = {
+    lossHistoryTimeWeights.map(_.asJava)
+  }
+
+  def getLossHistoryUnitWeightsJava: Option[util.List[Double]] = {
+    lossHistoryUnitWeights.map(_.asJava)
+  }
+}
+
+class DiffInDiffModel(override val uid: String)
+  extends Model[DiffInDiffModel]
+    with HasUnitCol
+    with HasTimeCol
+    with Wrappable
+    with ComplexParamsWritable
+    with SynapseMLLogging {
+
+  logClass(FeatureNames.Causal)
+
+  final val timeIndex = new DataFrameParam(this, "timeIndex", "time index")
+  def getTimeIndex: DataFrame = $(timeIndex)
+  def setTimeIndex(value: DataFrame): this.type = set(timeIndex, value)
+
+  final val timeIndexCol = new Param[String](this, "timeIndexCol", "time index column")
+  def getTimeIndexCol: String = $(timeIndexCol)
+  def setTimeIndexCol(value: String): this.type = set(timeIndexCol, value)
+
+  final val unitIndex = new DataFrameParam(this, "unitIndex", "unit index")
+  def getUnitIndex: DataFrame = $(unitIndex)
+  def setUnitIndex(value: DataFrame): this.type = set(unitIndex, value)
+
+  final val unitIndexCol = new Param[String](this, "unitIndexCol", "unit index column")
+  def getUnitIndexCol: String = $(unitIndexCol)
+  def setUnitIndexCol(value: String): this.type = set(unitIndexCol, value)
+
+  override protected lazy val pyInternalWrapper = true
+
+  def this() = this(Identifiable.randomUID("did"))
+
+  private final var summary: Option[DiffInDiffSummary] = None
+
+  def getSummary: DiffInDiffSummary = summary.getOrElse {
+    throw new SparkException(
+      s"No summary available for this ${this.getClass.getSimpleName}")
+  }
+
+  private[causal] def setSummary(summary: Option[DiffInDiffSummary]): this.type = {
+    this.summary = summary
+    this
+  }
+
+  override def copy(extra: ParamMap): DiffInDiffModel = {
+    copyValues(new DiffInDiffModel(uid))
```

Comment:
you dont seem to handle extra, you can use defaultcopy i think

Reply:
Good catch! I'm now passing `extra` to `copyValues`. The effect is same as `defaultCopy`.